### ISSUE TITLE
Ensure single instance of GazetteerCache is shared across Gunicorn workers

### ISSUE THREAD

**Issue by hectcastro** [open-apparel-registry/open-apparel-registry#1085]
With synchronous workers enabled, it is believed that Gunicorn spins up an instance of the application for each worker. For the [`GazeteerCache`](https://github.com/open-apparel-registry/open-apparel-registry/commit/5f0fb4b974741803eed08a9d2c9367ada589e6d9), it would be beneficial to load up a single instance of the cache, and share is amongst the Gunicorn workers.

This issue is to determine if the `GazeteerCache` is being instantiated multiple times per Gunicorn instance, and if so, try to reduce it to 1 instance per Gunicorn instance.

See: https://docs.gunicorn.org/en/stable/settings.html#preload-app

**Comment by rbreslow** [comment]
The [`Arbiter`](https://github.com/benoitc/gunicorn/blob/e636bf81989bb833d2b99104feb11e86c3f2c43a/gunicorn/arbiter.py) is the Gunicorn class that maintains the worker processes. Workers are [spawned](https://github.com/benoitc/gunicorn/blob/e636bf81989bb833d2b99104feb11e86c3f2c43a/gunicorn/arbiter.py#L567) using the [`fork()`](https://www.man7.org/linux/man-pages/man2/fork.2.html) syscall. 

> `fork()` creates a new process by duplicating the calling process. The new process is referred to as the _child_ process.  The calling process is referred to as the _parent_ process. 

> The child process and the parent process run in separate memory spaces.  At the time of `fork()` both memory spaces have the same content.

> On success, the PID of the child process is returned in the parent, and `0` is returned in the child.

The child process resumes at exactly the same place as the parent process.  If `fork()` returns a non-zero PID, the process continues and assumes the role of the **master** process, which manages the Gunicorn workers.

If `fork()` returns `0`, the process continues and assumes the role of the **child** process.  [`init_process()`](https://github.com/benoitc/gunicorn/blob/e636bf81989bb833d2b99104feb11e86c3f2c43a/gunicorn/workers/base.py#L85) is invoked which [imports](https://github.com/benoitc/gunicorn/blob/5425af8941e1d43c57b5b2fb3fbd7f728688ce01/gunicorn/workers/base.py#L134) [the](https://github.com/benoitc/gunicorn/blob/5425af8941e1d43c57b5b2fb3fbd7f728688ce01/gunicorn/workers/base.py#L144-L149) [WSGI](https://github.com/benoitc/gunicorn/blob/e636bf81989bb833d2b99104feb11e86c3f2c43a/gunicorn/app/wsgiapp.py#L38-L39) [app](https://github.com/benoitc/gunicorn/blob/7d8c92f48a6d9cee6b15fbdade6981721182d073/gunicorn/util.py#L351) and then begins [the event loop](https://github.com/benoitc/gunicorn/blob/5425af8941e1d43c57b5b2fb3fbd7f728688ce01/gunicorn/workers/sync.py#L59-L86) that [passes requests](https://github.com/benoitc/gunicorn/blob/5425af8941e1d43c57b5b2fb3fbd7f728688ce01/gunicorn/workers/sync.py#L177) to Django's [WSGI interface](https://github.com/django/django/blob/41a3b3d18647b258331104520e76f977406c590d/django/core/handlers/wsgi.py#L129-L149). 

Specifically, `util.import_app` will load our WSGI module (`oar.wsgi` in the command line argv) and try to [find](https://github.com/benoitc/gunicorn/blob/7d8c92f48a6d9cee6b15fbdade6981721182d073/gunicorn/util.py#L354) and [invoke](https://github.com/benoitc/gunicorn/blob/7d8c92f48a6d9cee6b15fbdade6981721182d073/gunicorn/util.py#L412) the method named `application`. 

https://github.com/open-apparel-registry/open-apparel-registry/blob/2f97c8f0b1075d844c935b202953ac1bcc3ff4a7/src/django/oar/wsgi.py#L16

When `application` is invoked, [`get_wsgi_application()`](https://github.com/django/django/blob/41a3b3d18647b258331104520e76f977406c590d/django/core/wsgi.py#L5-L13) is invoked, which invokes [`django.setup()`](https://github.com/django/django/blob/41a3b3d18647b258331104520e76f977406c590d/django/__init__.py#L8-L24), which loads the `settings` module in order to populate the app registry.

The important thing is that the `urls` module, where `GazetteerCache` [is first initialized](https://github.com/open-apparel-registry/open-apparel-registry/blob/2f97c8f0b1075d844c935b202953ac1bcc3ff4a7/src/django/oar/urls.py#L30), is not [loaded](https://github.com/django/django/blob/54dcfbc36780054d39299bfd1078938e5dd4e839/django/urls/resolvers.py#L575-L579) until the first request comes through the [resolver](https://github.com/django/django/blob/ecaac9e42f497be04ddc72dfebb6e397ccca9517/django/core/handlers/base.py#L275).

So, to take a step back, the `GazetteerCache` isn't instantiated until we're already running in the child process which runs in its own memory space. It cannot be shared with the other Gunicorn processes.

But what about [preloading](https://docs.gunicorn.org/en/stable/settings.html#preload-app)? When preloading is enabled, the arbiter will load the WSGI module before the worker processes are forked, following the flow I described above with `util.import_app`.

Because `GazetteerCache` is instantiated upon the first request, not when the WSGI module is loaded, this still doesn't get us anywhere. 

Now, you could instantiate the `GazetteerCache` at the top of `oar.wsgi`, and this would cause it to be started in the master parent process. However, the child process is created with a single thread—the one that called `fork()`. And, the entire virtual address space of the parent is replicated in the child, including the states of mutexes, etc.  This means the [`GazetteerCache._lock`](https://github.com/open-apparel-registry/open-apparel-registry/blob/2f97c8f0b1075d844c935b202953ac1bcc3ff4a7/src/django/api/matching.py#L429) will remain acquired in perpetuity.

This is where I realized we were thinking about this all wrong. The threading factor here is only used to initialize the cache (which takes a while) as to not block the first request to the app. As soon as that's done, the thread dies, and the rest of the interactions with `GazetteerCache` happen within the main thread. And even so, there's no way to share a thread with another process, without some form of [IPC](https://docs.python.org/3/library/ipc.html).

So, as the cache is architected, there's no way to share it across the Gunicorn workers. I can recommend a couple paths forward:

- Instantiate the cache in the master process using a [manager object](https://docs.python.org/3/library/multiprocessing.html#managers). Have the workers communicate with the master process via a [proxy](https://docs.python.org/3/library/multiprocessing.html#proxy-objects).
- Use a cache like Redis or memcached.

---

**TL;DR** 

Each worker is a unique instance of the application. As the cache is currently architected, there is no way to share it across workers. Assuming "Gunicorn instance" refers to the master process and all the workers, then yes the cache is instantiated for every worker.

**Comment by jwalgran** [comment]
Thank you for the in-depth analysis.

Regarding paths forward, the trained model which we ideally want to share between our workers, is an instance of the `dedupe.Gazetteer` class which currently takes up multiple gigabytes of RAM in production. I don't know of a way in which a cache server like Redis or memcached could help in this situation. Some form of IPC seems more appropriate. Managers look like a good place to start, however we would need to do a bit of refactoring since we would no longer be able to pass around a reference to the `dedupe.Gazetteer` instance. Training the model on startup requires querying Django model data, so whatever process that "holds" the shared instance will need to have the Django environment available.

**Comment by rbreslow** [comment]
> Regarding paths forward, the trained model which we ideally want to share between our workers, is an instance of the `dedupe.Gazetteer` class which currently takes up multiple gigabytes of RAM in production. I don't know of a way in which a cache server like Redis or memcached could help in this situation.

I don't know anything about the underlying architecture of the `deduce.Gazetteer` class, so you may be right. My intuition was that we have large maps of data that multiple processes need access to, and if that's the case, something like Redis stood out as a tool that can handle high volumes of this data with a high throughput.

With each approach I mentioned there'd be refactoring and rethinking required in order to take advantage of the new technology.

I agree that managers seem like they'd be more of a "drop in" approach as we could retain the concept of a single object.

> Training the model on startup requires querying Django model data, so whatever process that "holds" the shared instance will need to have the Django environment available.

If we initialize the shared instance in the Gunicorn master process, it would have access to the Django environment and be able to interact with the database.

By the way, I'm out of the loop on the performance issues that triggered this investigation. Do you want to get together to chat about next steps?

**Comment by hectcastro** [comment]
This is an interesting writeup of the inner details of the Gunicorn server model, especially as it relates to Django. I'm going to try and keep my summary comments above the horizontal rule, but have other comments inline underneath it.

> By the way, I'm out of the loop on the performance issues that triggered this investigation. Do you want to get together to chat about next steps?

I think most of what this issue intends to cover was spawned by the challenges https://github.com/open-apparel-registry/open-apparel-registry/issues/1084 is trying to solve for (which are relatively unrelated to the challenges discussed here). The specific focus on `GazetteerCache` is to gain a better understanding of application resource utilization _before_ we go down the path of trying to solve problems by vertically scaling resources.

Justin can correct me if I'm wrong, but I don't think a significant refactoring of the `GazetteerCache` access model is in the cards right now (the hope of `preload-app` was that enabling it and moving `GazetteerCache` around could cut resource consumption down by a lot). And, based on our collective understanding, adoption of a shared memory cache seems relatively intractable.

Given that, my takeaway from these findings are that we should consider reducing the Gunicorn worker count to one (1) and instantiate more instances of the ECS service. While conceptually that's similar to having one Gunicorn instance with multiple web workers, from a resource management perspective, it seems easier to make decisions when there is a clear 1:1 mapping.

---

> Now, you could instantiate the `GazetteerCache` at the top of `oar.wsgi`, and this would cause it to be started in the master parent process. However, the child process is created with a single thread—the one that called `fork()`. And, the entire virtual address space of the parent is replicated in the child, including the states of mutexes, etc.  This means the [`GazetteerCache._lock`](https://github.com/open-apparel-registry/open-apparel-registry/blob/2f97c8f0b1075d844c935b202953ac1bcc3ff4a7/src/django/api/matching.py#L429) will remain acquired in perpetuity.

I think that the perpetual lock referenced here is an implementation detail unique to the initial pass at solving this problem (providing the application with an up-to-date model for doing comparisons). It seems safe to say that an approach geared toward working in a single, shared instance of the `GazetteerCache` (across multiple workers) would be reworked to meet the specifics of that model.

> This is where I realized we were thinking about this all wrong. The threading factor here is only used to initialize the cache (which takes a while) as to not block the first request to the app. As soon as that's done, the thread dies, and the rest of the interactions with `GazetteerCache` happen within the main thread. And even so, there's no way to share a thread with another process, without some form of [IPC](https://docs.python.org/3/library/ipc.html).

It is unclear what thinking was "wrong". I think the thread was used as a way to allow the initial request to succeed, and the use of the lock was intended to block access to the model (by other requests) until it was made up-to-date. I don't think the thread was intended to simulate multi-process IPC.

> If we initialize the shared instance in the Gunicorn master process, it would have access to the Django environment and be able to interact with the database.

This seems like a valuable piece of information if we pursued anything that looks like a shared `GazetteerCache` across processes.

**Comment by jwalgran** [comment]
>Justin can correct me if I'm wrong, but I don't think a significant refactoring of the GazetteerCache access model is in the cards right now.

That's correct. We have a limited amount of time to work on this until the 4th quarter.

**Comment by rbreslow** [comment]
> It seems safe to say that an approach geared toward working in a single, shared instance of the `GazetteerCache` (across multiple workers) would be reworked to meet the specifics of that model.

That's correct. However, without the threading, initializing the current version of `GazetteerCache` with preloading wouldn't really do much. It would block the rest of the workers until it finishes building which would result in a longer start up time with no ability to share the cache afterwards.

> It is unclear what thinking was "wrong". I think the thread was used as a way to allow the initial request to succeed, and the use of the lock was intended to block access to the model (by other requests) until it was made up-to-date. I don't think the thread was intended to simulate multi-process IPC.

When I went into this, I didn't understand the thread was for initializing the cache. I thought the cache lived in the thread and the goal was to try to share the thread between processes. When that detail was revealed, my thought process on this issue changed. Might change "we" to "I."

> Given that, my takeaway from these findings are that we should consider reducing the Gunicorn worker count to one (1) and instantiate more instances of the ECS service. While conceptually that's similar to having one Gunicorn instance with multiple web workers, from a resource management perspective, it seems easier to make decisions when there is a clear 1:1 mapping.

That sounds like a good idea. I created #1091.
